### PR TITLE
Altering DiracDeterminantBatched and friends to take Dual arguments.

### PR DESCRIPTION
**Pull request by PDoakORNL**
For portable accelerated Determinant inversion implementations APIs should reflect whether an argument is expected to be Dual (i.e. exist on host and accelerator) or not.

Also renaming of types and some member variables for clarity and readability.

Minimum change to existing implementation possible.

## Proposed changes
API argument changes.
Rename Refactorings

## What type(s) of changes does this code introduce?

- New feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Documentation changes

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
Leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by PDoakORNL** [comment]
start testing in house

**Review by ye-luo — CHANGES_REQUESTED** [review]
Reviewing this amount of change is still challenging but the change logics are clear and I don't get distracted by reg flags here and there like previous attempts.

No Major concerns and getting this in will be easy. So please keep address the bits from the review. Don't rebase or squash. Just add new commits and use merge.

Temporarily blocked as I would like to try some runs.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 1]
```diff
@@ -47,15 +52,23 @@ void DiracDeterminantBatched<DET_ENGINE>::set(int first, int nel, int delay)
 }
 
 template<typename DET_ENGINE>
-void DiracDeterminantBatched<DET_ENGINE>::invertPsiM(const Matrix<Value>& logdetT)
+void DiracDeterminantBatched<DET_ENGINE>::invertPsiM(const DualMatrix<Value>& psiM, DualMatrix<Value>& psiMinv)
 {
   ScopedTimer inverse_timer(InverseTimer);
-  det_engine_.invert_transpose(logdetT, log_value_);
+  det_engine_.invert_transpose(psiM, psiMinv, log_value());
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:58]: I should not have used this here.

get_log_value() is const
log_value() returns a non_const reference.
This is for API's where we must pass the log_value to the API as a return argument but that actually breaks encapsulation.
If more functions returned values then we could have a `set_log_value(LogValue log_value)`
Whether in this PR or later that was needed for compatibility with code outside of the feature branch.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 2]
```diff
@@ -47,15 +52,23 @@ void DiracDeterminantBatched<DET_ENGINE>::set(int first, int nel, int delay)
...
+
+#ifndef NDEBUG
+  // This is easily breakable in that it assumes this function gets psiMinv == det_engine_.psiMinv_
+  auto& engine_psiMinv = det_engine_.psiMinv();
+  dummy_vmt.attachReference(engine_psiMinv.data(), engine_psiMinv.rows(), engine_psiMinv.cols());
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:63]: Say you have a DiracDeterminantBase reference or pointer, as I have had numerous times in debug/testing you want to be able to inspect the psiMinv in a generic way even though where and how that is stored is implementation specific. calling the virtual method (debug only) getPsiMInv() gets you a reference to a CPU psiMInv.
The call in DiracDeterminantBase is getPsiMInv() instead of get_psiMinv to indicate that this isn't a practically zero cost accessor. I had many unit tests at one point that were more thorough comparing determinant implementations when in debug.  They're lost back who knows how many rejected PR's ago.  This is still useful.  But I go out of my way to make it unusable for anything but debug testing and debugging.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 3]
```diff
@@ -47,15 +52,23 @@ void DiracDeterminantBatched<DET_ENGINE>::set(int first, int nel, int delay)
...
+#endif
 }
 
 template<typename DET_ENGINE>
 void DiracDeterminantBatched<DET_ENGINE>::mw_invertPsiM(const RefVectorWithLeader<WaveFunctionComponent>& wfc_list,
-                                                        const RefVector<const Matrix<Value>>& logdetT_list)
+                                                        RefVector<const DualMatrix<Value>>& logdetT_list,
+                                                        RefVector<DualMatrix<Value>>& a_inv_list,
+                                                        const std::vector<bool>& compute_mask) const
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:71]: mw_invertPsiM doesn't have an override, maybe it did once?

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 4]
```diff
@@ -65,14 +78,31 @@ void DiracDeterminantBatched<DET_ENGINE>::mw_invertPsiM(const RefVectorWithLeade
...
+    det.log_value_ = wfc_leader.mw_res_->log_values[iw];
   }
 
-  DET_ENGINE::mw_invert_transpose(engine_list, logdetT_list, log_value_list);
+#ifndef NDEBUG
+  for (int iw = 0; iw < nw; ++iw)
+  {
+    auto& det            = wfc_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
+    auto& engine_psiMinv = det.get_det_engine().psiMinv();
+    det.dummy_vmt.attachReference(engine_psiMinv.data(), engine_psiMinv.rows(), engine_psiMinv.cols());
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:103]: has the same purpose as above.  Base class visibility of psiMinv.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 5]
```diff
@@ -85,7 +115,7 @@ void DiracDeterminantBatched<DET_ENGINE>::resize(int nel, int morb)
   psiM_vgl.resize(nel * norb);
   // attach pointers VGL
   psiM_temp.attachReference(psiM_vgl.data(0), nel, norb);
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:117]: yes good catch.  psiM_temp switched from host to dual and back several times as I got the API and minimal change working.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:63]: I support removing this whole apparatus once the inversion is in and stable.

File: src/QMCWaveFunctions/Fermion/DiracMatrix.h [hunk 6]
```diff
@@ -126,7 +134,7 @@ class DiracMatrix
       msg << "Xgetri failed with error " << status << std::endl;
       throw std::runtime_error(msg.str());
     }
-        
+
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrix.h:137]: I wanted to write docs at least for Xgetrf since I had to look that up. Those make much more sense if the overloads for each function are together.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 7]
```diff
@@ -145,29 +177,28 @@ void DiracDeterminantBatched<DET_ENGINE>::mw_evalGrad(const RefVectorWithLeader<
...
 
   {
     ScopedTimer local_timer(RatioTimer);
+    auto& psiMinv          = det_engine_.psiMinv();
     const int WorkingIndex = iat - FirstIndex;
     curRatio               = simd::dot(psiMinv[WorkingIndex], psiV.data(), NumOrbitals);
-    grad_iat += static_cast<ValueType>(static_cast<PsiValueType>(1.0) / curRatio) *
-        simd::dot(psiMinv[WorkingIndex], dpsiV.data(), NumOrbitals);
+    grad_iat += static_cast<Value>(static_cast<PsiValue>(1.0) / curRatio) *
+        simd::dot(det_engine_.psiMinv()[WorkingIndex], dpsiV.data(), NumOrbitals);
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:197]: I think the issue here is the engine implementation may know whether psiMinv is consistent but the DDB shouldn't.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:197]: not yet but I've added comments explaining why.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 8]
```diff
@@ -266,13 +295,14 @@ void DiracDeterminantBatched<DET_ENGINE>::mw_accept_rejectMove(
...
-  std::vector<ValueType*> psiM_l_dev_ptr_list(n_accepted, nullptr);
+  std::vector<Value*> psiM_g_dev_ptr_list(n_accepted, nullptr);
+  std::vector<Value*> psiM_l_dev_ptr_list(n_accepted, nullptr);
 
   const int WorkingIndex = iat - FirstIndex;
   for (int iw = 0, count = 0; iw < nw; iw++)
   {
-    auto& det = wfc_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
+    // This can be auto but some debuggers can't figure the type out.
+    DiracDeterminantBatched<DET_ENGINE>& det = wfc_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:305]: Me too but debugging is also important.  Can be changed to auto when feature is stablized.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h [hunk 9]
```diff
@@ -175,7 +181,8 @@ class DiracDeterminantBatched : public DiracDeterminantBase
                       const RefVector<ParticleSet::ParticleGradient_t>& G_list,
                       const RefVector<ParticleSet::ParticleLaplacian_t>& L_list) const override;
 
-  void recompute(const ParticleSet& P) override;
+  // void recompute( //DiracDeterminantBatchedMultiWalkerResource& mw_res,
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h:184]: removed

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h [hunk 10]
```diff
@@ -236,28 +243,36 @@ class DiracDeterminantBatched : public DiracDeterminantBase
...
+  Vector<Value> d2psiV_host_view;
 
-  // psi(r')/psi(r) during a PbyP move
+  /// psi(r')/psi(r) during a PbyP move
   PsiValue curRatio;
+  /**@}*/
 
   std::unique_ptr<DiracDeterminantBatchedMultiWalkerResource> mw_res_;
 
+  void recompute(const ParticleSet& P) override;
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h:259]: done

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h [hunk 11]
```diff
@@ -207,21 +214,21 @@ class DiracDeterminantBatched : public DiracDeterminantBase
...
    */
-  Matrix<Value> psiMinv;
-
   /// memory for psiM, dpsiM and d2psiM. [5][norb*norb]
   DualVGLVector psiM_vgl;
   /// psiM(j,i) \f$= \psi_j({\bf r}_i)\f$. partial memory view of psiM_vgl
-  Matrix<Value> psiM_temp;
+  DualMatrix<Value> psiM_temp;
   /// dpsiM(i,j) \f$= \nabla_i \psi_j({\bf r}_i)\f$. partial memory view of psiM_vgl
   Matrix<Grad> dpsiM;
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h:233]: Expanded.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp [hunk 12]
```diff
@@ -43,29 +43,33 @@ namespace qmcplusplus
...
+  using HandleResource = DummyResource;
+
+private:
   aligned_vector<VALUE_FP> m_work_;
   int lwork_;
 
   /// Matrices held in memory matrices n^2 * nw  elements
   OffloadPinnedVector<VALUE_FP> psiM_fp_;
-
+  OffloadPinnedVector<VALUE_FP> logdets_fp_;
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp:68]: gone

File: src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h [hunk 13]
```diff
@@ -16,85 +16,101 @@
...
+  template<typename DT>
+  using UnpinnedDualAllocator = OffloadAllocator<DT>;
   template<typename DT>
   using PinnedDualAllocator = OffloadPinnedAllocator<DT>;
 
-  using OffloadValueVector_t       = Vector<T, OffloadAllocator<T>>;
-  using OffloadPinnedValueVector_t = Vector<T, OffloadPinnedAllocator<T>>;
-  using OffloadPinnedValueMatrix_t = Matrix<T, OffloadPinnedAllocator<T>>;
+  template<typename DT>
+  using UnpinnedDualVector = Vector<DT, UnpinnedDualAllocator<DT>>;
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h:51]: Fixed this, I was on the fence about it.  PinnedDualAllocator still needs to be defined for DiracBatchedDeterminant it still needs knowledge of the PinnedDualAllocator type.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h:51]: Correct PinnedDualAllocator is intended to communicate with the outside.

File: src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp [hunk 14]
```diff
@@ -166,6 +171,13 @@ void testTrialWaveFunction_diamondC_2x1x1(const int ndelay)
   RadialJastrowBuilder jb(c, elec_);
   psi.addComponent(jb.buildComponent(jas1));
 
+  // testing batched interfaces
+  ResourceCollection pset_res("test_pset_res");
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/tests/test_TrialWaveFunction_diamondC_2x1x1.cpp:175]: moved. But I don't think we should support the legacy driver at all from DiracDeterminantBatched and this doesn't show correct usage.  Someone now has to span several hundred line to intuit correct use of the object.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h [hunk 15]
```diff
@@ -207,21 +214,21 @@ class DiracDeterminantBatched : public DiracDeterminantBase
...
 
-  /// return  for testing
-  auto& getPsiMinv() const { return psiMinv; }
+  DET_ENGINE& get_det_engine() { return det_engine_; }
 
-  /** inverse transpose of psiM(j,i) \f$= \psi_j({\bf r}_i)\f$ memory view
-   * Actual memory is owned by det_engine_
-   * Only NumOrbitals x NumOrbitals subblock has meaningful data
-   * The number of rows is equal to NumOrbitals
-   * The number of columns in each row is padded to a multiple of QMC_SIMD_ALIGNMENT
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h:217]: Removed

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp]: I change the name of the non const reference return to get_ref_log_value

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 16]
```diff
@@ -381,15 +413,18 @@ typename DiracDeterminantBatched<DET_ENGINE>::LogValue DiracDeterminantBatched<D
...
+    // this updates LogValue
     evaluateLog(P, G, L);
   else
   {
     if (UpdateMode == ORB_PBYP_RATIO)
     { //need to compute dpsiM and d2psiM. Do not touch psiM!
       ScopedTimer spo_timer(SPOVGLTimer);
-      Phi->evaluate_notranspose(P, FirstIndex, LastIndex, psiM_temp, dpsiM, d2psiM);
+      Matrix<Value> psiM_temp_host(psiM_temp.data(), psiM_temp.rows(), psiM_temp.cols());
+      Matrix<Grad> dpsiM_host(dpsiM.data(), dpsiM.rows(), dpsiM.cols());
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp]: fixed

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.h:210]: I probably misread this change. github source code alignment is misleading. I mean the information of padding is lost.

File: src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h [hunk 17]
```diff
@@ -16,85 +16,99 @@
...
+  DiracMatrix<VALUE_FP> detEng;
+
+  int inv_row_id_;
+
+  /** matrix inversion engine this crowd scope resouce and only the leader engine gets it.
+   *  With some care this could be the DiracMatrixComputeCUDA 
+   *  as well as DiracMatrixComputeOMPTarget
+   */
+  using DetInverter = DiracMatrixComputeOMPTarget<VALUE_FP>;
+  UPtr<DetInverter> det_inverter_;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h:90]: This doesn't seem to belong to this PR and I cannot find det_inverter_ being used.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 18]
```diff
@@ -899,16 +947,20 @@ void DiracDeterminantBatched<DET_ENGINE>::mw_recompute(const RefVectorWithLeader
...
     {
       auto& det          = wfc_filtered_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
       auto* psiM_vgl_ptr = det.psiM_vgl.data();
       size_t stride      = wfc_leader.psiM_vgl.capacity();
       PRAGMA_OFFLOAD("omp target update to(psiM_vgl_ptr[stride:stride*4]) nowait")
       const_psiM_temp_list.push_back(det.psiM_temp);
+      psiMinv_list.push_back(det.get_det_engine().psiMinv());
     }
-    mw_invertPsiM(wfc_filtered_list, const_psiM_temp_list);
+    auto& mw_res = *wfc_leader.mw_res_;
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:923]: I don't see mw_res being used.

**Review by ye-luo — COMMENTED** [review]
Some remaining issues.

**Comment by ye-luo** [comment]
I will directly take care of my latest review comments. Doing testing now.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp]: 0e048da removes the unsafe reference accessor.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:51]: defer to #3461

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:90]: defer to #3461

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp]: DDB has a certain assumption which side should be update.

File: src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h (hunk 17 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h:90]: It is needed for now.

**Comment by ye-luo** [comment]
Start testing in-house

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Forgot to say. The unnecessary challenge added by renaming is largely avoidable if separate PRs are made. It was my fault saying OK to incorporate renaming in this PR when I requested it being split from the original PR.
